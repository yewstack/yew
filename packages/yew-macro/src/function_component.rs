use proc_macro2::TokenStream;
use quote::{quote, quote_spanned, ToTokens};
use syn::parse::{Parse, ParseStream};
use syn::punctuated::Punctuated;
use syn::spanned::Spanned;
use syn::token::Comma;
use syn::{Attribute, Block, FnArg, Generics, Ident, Item, ItemFn, ReturnType, Type, Visibility};

pub struct FunctionComponent {
    block: Box<Block>,
    props_type: Box<Type>,
    arg: FnArg,
    generics: Generics,
    vis: Visibility,
    attrs: Vec<Attribute>,
    name: Ident,
    return_type: Box<Type>,
}

impl Parse for FunctionComponent {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let parsed: Item = input.parse()?;

        match parsed {
            Item::Fn(func) => {
                let ItemFn {
                    attrs,
                    vis,
                    sig,
                    block,
                } = func;

                if sig.generics.lifetimes().next().is_some() {
                    return Err(syn::Error::new_spanned(
                        sig.generics,
                        "function components can't have generic lifetime parameters",
                    ));
                }

                if sig.asyncness.is_some() {
                    return Err(syn::Error::new_spanned(
                        sig.asyncness,
                        "function components can't be async",
                    ));
                }

                if sig.constness.is_some() {
                    return Err(syn::Error::new_spanned(
                        sig.constness,
                        "const functions can't be function components",
                    ));
                }

                if sig.abi.is_some() {
                    return Err(syn::Error::new_spanned(
                        sig.abi,
                        "extern functions can't be function components",
                    ));
                }

                let return_type = match sig.output {
                    ReturnType::Default => {
                        return Err(syn::Error::new_spanned(
                            sig,
                            "function components must return `yew::Html`",
                        ))
                    }
                    ReturnType::Type(_, ty) => ty,
                };

                let mut inputs = sig.inputs.into_iter();
                let arg: FnArg = inputs
                    .next()
                    .unwrap_or_else(|| syn::parse_quote! { _: &() });

                let ty = match &arg {
                    FnArg::Typed(arg) => match &*arg.ty {
                        Type::Reference(ty) => {
                            if ty.lifetime.is_some() {
                                return Err(syn::Error::new_spanned(
                                    &ty.lifetime,
                                    "reference must not have a lifetime",
                                ));
                            }

                            if ty.mutability.is_some() {
                                return Err(syn::Error::new_spanned(
                                    &ty.mutability,
                                    "reference must not be mutable",
                                ));
                            }

                            ty.elem.clone()
                        }
                        ty => {
                            let msg = format!(
                                "expected a reference to a `Properties` type (try: `&{}`)",
                                ty.to_token_stream()
                            );
                            return Err(syn::Error::new_spanned(ty, msg));
                        }
                    },

                    FnArg::Receiver(_) => {
                        return Err(syn::Error::new_spanned(
                            arg,
                            "function components can't accept a receiver",
                        ));
                    }
                };

                // Checking after param parsing may make it a little inefficient
                // but that's a requirement for better error messages in case of receivers
                // `>0` because first one is already consumed.
                if inputs.len() > 0 {
                    let params: TokenStream = inputs.map(|it| it.to_token_stream()).collect();
                    return Err(syn::Error::new_spanned(
                        params,
                        "function components can accept at most one parameter for the props",
                    ));
                }

                Ok(Self {
                    props_type: ty,
                    block,
                    arg,
                    generics: sig.generics,
                    vis,
                    attrs,
                    name: sig.ident,
                    return_type,
                })
            }
            item => Err(syn::Error::new_spanned(
                item,
                "`function_component` attribute can only be applied to functions",
            )),
        }
    }
}

pub struct FunctionComponentName {
    component_name: Ident,
}

impl Parse for FunctionComponentName {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        if input.is_empty() {
            return Err(input.error("expected identifier for the component"));
        }

        let component_name = input.parse()?;

        Ok(Self { component_name })
    }
}

pub fn function_component_impl(
    name: FunctionComponentName,
    component: FunctionComponent,
) -> syn::Result<TokenStream> {
    let FunctionComponentName { component_name } = name;

    let FunctionComponent {
        block,
        props_type,
        arg,
        generics,
        vis,
        attrs,
        name: function_name,
        return_type,
    } = component;

    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();

    if function_name == component_name {
        return Err(syn::Error::new_spanned(
            component_name,
            "the component must not have the same name as the function",
        ));
    }

    let ret_type = quote_spanned!(return_type.span()=> ::yew::html::Html);

    let phantom_generics = generics
        .type_params()
        .map(|ty_param| ty_param.ident.clone()) // create a new Punctuated sequence without any type bounds
        .collect::<Punctuated<_, Comma>>();

    let quoted = quote! {
        #[doc(hidden)]
        #[allow(non_camel_case_types)]
        #[allow(unused_parens)]
        #vis struct #function_name #impl_generics {
            _marker: ::std::marker::PhantomData<(#phantom_generics)>,
        }

        impl #impl_generics ::yew::functional::FunctionProvider for #function_name #ty_generics #where_clause {
            type TProps = #props_type;

            fn run(#arg) -> #ret_type {
                #block
            }
        }

        #(#attrs)*
        #[allow(type_alias_bounds)]
        #vis type #component_name #impl_generics = ::yew::functional::FunctionComponent<#function_name #ty_generics>;
    };

    Ok(quoted)
}
