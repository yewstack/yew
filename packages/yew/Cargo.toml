[package]
name = "yew"
version = "0.17.4"
edition = "2018"
authors = [
  "Denis Kolodin <deniskolodin@gmail.com>",
  "Justin Starry <justin@yew.rs>",
]
repository = "https://github.com/yewstack/yew"
homepage = "https://github.com/yewstack/yew"
documentation = "https://docs.rs/yew/"
license = "MIT/Apache-2.0"
readme = "../README.md"
keywords = ["web", "webasm", "javascript"]
categories = ["gui", "wasm", "web-programming"]
description = "A framework for making client-side single-page apps"

[dependencies]
anyhow = "1"
anymap = "0.12"
console_error_panic_hook = "0.1"
gloo = "0.2.1"
http = "0.2"
indexmap = { version = "1.5", features = ["std"] }
js-sys = "0.3"
log = "0.4"
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
slab = "0.4"
thiserror = "1"
wasm-bindgen = "0.2.60"
wasm-bindgen-futures = "0.4"
yew-macro = { version = "0.17.0", path = "../yew-macro" }

# optional encodings
bincode = { version = "1", optional = true }
rmp-serde = { version = "0.15.0", optional = true }
serde_cbor = { version = "0.11.1", optional = true }
serde_yaml = { version = "0.8.4", optional = true }
toml = { version = "0.5", optional = true }

[dependencies.web-sys]
version = "0.3"
features = [
  "AbortController",
  "AbortSignal",
  "AnimationEvent",
  "BinaryType",
  "Blob",
  "BlobPropertyBag",
  "console",
  "DedicatedWorkerGlobalScope",
  "Document",
  "DomTokenList",
  "DragEvent",
  "Element",
  "ErrorEvent",
  "Event",
  "EventTarget",
  "File",
  "FileList",
  "FileReader",
  "FocusEvent",
  "Headers",
  "HtmlElement",
  "HtmlButtonElement",
  "HtmlInputElement",
  "HtmlSelectElement",
  "HtmlTextAreaElement",
  "InputEvent",
  "KeyboardEvent",
  "Location",
  "MessageEvent",
  "MouseEvent",
  "Node",
  "ObserverCallback",
  "PointerEvent",
  "ProgressEvent",
  "ReferrerPolicy",
  "Request",
  "RequestCache",
  "RequestCredentials",
  "RequestInit",
  "RequestMode",
  "RequestRedirect",
  "Response",
  "Storage",
  "Text",
  "TouchEvent",
  "TransitionEvent",
  "UiEvent",
  "Url",
  "WebSocket",
  "WheelEvent",
  "Window",
  "Worker",
  "WorkerGlobalScope",
  "WorkerOptions",
]

[dev-dependencies]
base64 = "0.13.0"
bincode = "1"
easybench-wasm = "0.2.1"
rmp-serde = "0.15.0"
rustversion = "1.0"
serde_derive = "1"
ssri = "6.0.0"
trybuild = "1.0"
wasm-bindgen-test = "0.3.4"

[features]
default = ["agent"]
doc_test = []
wasm_test = []
wasm_bench = []
agent = ["bincode"]
yaml = ["serde_yaml"]
msgpack = ["rmp-serde"]
cbor = ["serde_cbor"]

[package.metadata.docs.rs]
features = ["yaml", "cbor", "toml", "msgpack", "doc_test"]
