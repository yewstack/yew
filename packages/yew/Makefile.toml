[env]
CARGO_MAKE_CLIPPY_ARGS = "--features cbor,msgpack,toml,yaml -- --deny=warnings"

[tasks.test]
extend = "core::wasm-pack-base"
command = "wasm-pack"
args = [
    "test",
    "@@split(YEW_TEST_FLAGS, )",
    "--",
    "--features",
    "${YEW_TEST_FEATURES}",
]

[tasks.doc-test]
clear = true
run_task = { name = ["doc-test-normal"], fork = true }

[tasks.doc-test-normal]
command = "cargo"
args = [
    "test",
    "--doc",
    "--features",
    "doc_test,wasm_test,yaml,msgpack,cbor,toml",
]

[tasks.bench]
extend = "core::wasm-pack-base"
command = "wasm-pack"
args = [
    "test",
    "--release",
    "--firefox",
    "--headless",
    "--",
    "--features",
    "wasm_bench",
    "bench",
]
