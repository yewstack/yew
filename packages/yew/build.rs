use std::env;

pub fn main() {
    let using_web_sys = cfg!(feature = "web_sys");
    let using_std_web = cfg!(feature = "std_web");
    if using_web_sys && using_std_web {
        panic!("Yew does not allow the `web_sys` and `std_web` cargo features to be used simultaneously");
    } else if !using_web_sys && !using_std_web {
        panic!("Yew requires selecting either the `web_sys` or `std_web` cargo feature");
    }

    let using_cargo_web = env::var("COMPILING_UNDER_CARGO_WEB").is_ok();
    if using_cargo_web && using_web_sys {
        panic!("cargo-web is not compatible with web-sys");
    }
}
