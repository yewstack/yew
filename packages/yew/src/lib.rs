#![allow(clippy::needless_doctest_main)]
#![doc(html_logo_url = "https://static.yew.rs/logo.svg")]

//! # Yew Framework - API Documentation
//!
//! Yew is a modern Rust framework for creating multi-threaded front-end web apps using WebAssembly
//!
//! - Features a macro for declaring interactive HTML with Rust expressions. Developers who have experience using JSX in React should feel quite at home when using Yew.
//! - Achieves high performance by minimizing DOM API calls for each page render and by making it easy to offload processing to background web workers.
//! - Supports JavaScript interoperability, allowing developers to leverage NPM packages and integrate with existing JavaScript applications.
//!
//! ### Supported Targets
//! - `wasm32-unknown-unknown`
//!
//! ### Important Notes
//! - Yew is not (yet) production ready but is great for side projects and internal tools
#![cfg_attr(
    feature = "web_sys",
    doc = " - If your app is built with `stdweb`, we recommend using [`yew-stdweb`](https://docs.rs/yew-stdweb) instead."
)]
//!
//! ## Example
//!
//! ```rust
//! use yew::prelude::*;
//!
//! enum Msg {
//!     AddOne,
//! }
//!
//! struct Model {
//!     link: ComponentLink<Self>,
//!     value: i64,
//! }
//!
//! impl Component for Model {
//!     type Message = Msg;
//!     type Properties = ();
//!
//!     fn create(_props: Self::Properties, link: ComponentLink<Self>) -> Self {
//!         Self {
//!             link,
//!             value: 0,
//!         }
//!     }
//!
//!     fn update(&mut self, msg: Self::Message) -> ShouldRender {
//!         match msg {
//!             Msg::AddOne => {
//!                 self.value += 1;
//!                 true
//!             }
//!         }
//!     }
//!
//!     fn change(&mut self, _props: Self::Properties) -> ShouldRender {
//!         false
//!     }
//!
//!     fn view(&self) -> Html {
//!         html! {
//!             <div>
//!                 <button onclick=self.link.callback(|_| Msg::AddOne)>{ "+1" }</button>
//!                 <p>{ self.value }</p>
//!             </div>
//!         }
//!     }
//! }
//!
//!# fn dont_execute() {
//! fn main() {
//!     yew::start_app::<Model>();
//! }
//!# }
//! ```
//!

#![deny(
    missing_docs,
    missing_debug_implementations,
    bare_trait_objects,
    anonymous_parameters,
    elided_lifetimes_in_paths
)]
#![allow(macro_expanded_macro_exports_accessed_by_absolute_paths)]
#![recursion_limit = "512"]
extern crate self as yew;

/// This macro provides a convenient way to create [`Classes`].
///
/// The macro takes a list of items similar to the [`vec!`] macro and returns a [`Classes`] instance.
/// Each item can be of any type that implements `Into<Classes>` (See the implementations on [`Classes`] to learn what types can be used).
///
/// # Example
///
/// ```
/// # use yew::prelude::*;
/// # fn test() {
/// let conditional_class = Some("my-other-class");
/// let vec_of_classes = vec!["one-bean", "two-beans", "three-beans", "a-very-small-casserole"];
///
/// html! {
///     <div class=classes!("my-container-class", conditional_class, vec_of_classes)>
///         // ...
///     </div>
/// };
/// # }
/// ```
pub use yew_macro::classes;

/// This macro implements JSX-like templates.
///
/// This macro always returns [`Html`].
/// If you need to preserve the type of a component, use the [`html_nested!`] macro instead.
///
/// More information about using the `html!` macro can be found in the [Yew Docs]
///
/// [`Html`]: ./html/type.Html.html
/// [`html_nested!`]: ./macro.html_nested.html
/// [Yew Docs]: https://yew.rs/docs/en/concepts/html/
pub use yew_macro::html;

/// This macro is similar to [`html!`], but preserves the component type instead
/// of wrapping it in [`Html`].
///
/// That macro is useful when, for example, in a typical implementation of a list
/// component (let's assume it's called `List`).
/// In a typical implementation you might find two component types -- `List` and `ListItem`.
/// Only `ListItem` components are allowed to be children of List`.
///
/// You can find an example implementation of this in the [`nested_list`] example.
/// That example shows, how to create static lists with their children.
///
/// ```
/// # use yew::prelude::*;
/// use yew::html::ChildrenRenderer;
/// use yew::virtual_dom::VChild;
///
/// #[derive(Clone, Properties)]
/// struct List {
///   children: ChildrenRenderer<ListItem>,
/// }
/// impl Component for List {
/// #   type Message = ();
///   type Properties = Self;
///   // ...
/// #   fn create(props: Self::Properties, _: ComponentLink<Self>) -> Self { props }
/// #   fn update(&mut self, _: Self::Message) -> ShouldRender { false }
/// #   fn change(&mut self, _: Self::Properties) -> ShouldRender { false }
/// #   fn view(&self) -> Html { unimplemented!() }
/// }
///
/// #[derive(Clone)]
/// struct ListItem;
/// impl Component for ListItem {
/// #   type Message = ();
/// #   type Properties = ();
///   // ...
/// #   fn create(_: Self::Properties, _: ComponentLink<Self>) -> Self { Self }
/// #   fn update(&mut self, _: Self::Message) -> ShouldRender { false }
/// #   fn change(&mut self, _: Self::Properties) -> ShouldRender { false }
/// #   fn view(&self) -> Html { unimplemented!() }
/// }
///
/// // Required for ChildrenRenderer
/// impl From<VChild<ListItem>> for ListItem {
///   fn from(child: VChild<ListItem>) -> Self { Self }
/// }
///
/// impl Into<Html> for ListItem {
///   fn into(self) -> Html { self.view() }
/// }
///
/// // You can use `List` with nested `ListItem` components.
/// // Using any other kind of element would result in a compile error.
/// # fn test() -> Html {
/// html! {
///   <List>
///     <ListItem/>
///     <ListItem/>
///     <ListItem/>
///   </List>
/// }
/// # }
/// # fn test_iter() -> Html {
/// # let some_iter = (0..10);
/// // In many cases you might want to create the content dynamically.
/// // To do this, you can use the following code:
/// html! {
///   <List>
///     { for some_iter.map(|_| html_nested!{ <ListItem/> }) }
///   </List>
/// }
/// # }
/// ```
///
/// If you used the [`html!`] macro instead of `html_nested!`, the code would
/// not compile because we explicitly indicated to the compiler that `List`
/// can only contain elements of type `ListItem` using [`ChildrenRenderer<ListItem>`],
/// while [`html!`] creates items of type [`Html`].
///
///
/// [`html!`]: ./macro.html.html
/// [`Html`]: ./html/type.Html.html
/// [`nested_list`]: https://github.com/yewstack/yew/tree/master/examples/nested_list
/// [`ChildrenRenderer<ListItem>`]: ./html/struct.ChildrenRenderer.html
pub use yew_macro::html_nested;

/// Build [`Properties`] outside of the [`html!`] macro.
///
/// It's already possible to create properties like normal Rust structs
/// but if there are lots of optional props the end result is often needlessly verbose.
/// This macro allows you to build properties the same way the [`html!`] macro does.
///
/// The macro doesn't support special props like `ref` and `key`, they need to be set in the [`html!`] macro.
///
/// You can read more about `Properties` in the [Yew Docs].
///
/// # Example
///
/// ```
/// # use yew::prelude::*;
/// use std::borrow::Cow;
///
/// #[derive(Clone, Properties)]
/// struct Props {
///     #[prop_or_default]
///     id: usize,
///     name: Cow<'static, str>,
/// }
///
/// struct Model(Props);
/// impl Component for Model {
/// #   type Message = ();
///     type Properties = Props;
///     // ...
/// #   fn create(_props: Self::Properties, _link: ComponentLink<Self>) -> Self { unimplemented!() }
/// #   fn update(&mut self, _msg: Self::Message) -> ShouldRender { unimplemented!() }
/// #   fn change(&mut self, _props: Self::Properties) -> ShouldRender { unimplemented!() }
/// #   fn view(&self) -> Html { unimplemented!() }
/// }
///
/// # fn foo() -> Html {
/// // You can build props directly ...
/// let props = yew::props!(Props { name: Cow::from("Minka") });
/// # assert_eq!(props.name, "Minka");
/// // ... or build the associated properties of a component
/// let props = yew::props!(Model::Properties { id: 2, name: Cow::from("Lemmy") });
/// # assert_eq!(props.id, 2);
///
/// // Use the `with props` syntax to create a component with the props.
/// html! {
///     <Model key=1 with props />
/// }
/// # }
/// ```
///
/// [`html!`]: ./macro.html.html
/// [`Properties`]: ./html/trait.Properties.html
/// [yew docs]: https://yew.rs/docs/en/concepts/components/properties
pub use yew_macro::props;

/// This module contains macros which implements html! macro and JSX-like templates
pub mod macros {
    pub use crate::classes;
    pub use crate::html;
    pub use crate::html_nested;
    pub use crate::props;
}

pub mod app;
pub mod callback;
pub mod format;
pub mod html;
mod scheduler;
pub mod utils;
pub mod virtual_dom;

#[cfg(feature = "agent")]
pub mod agent;

pub use web_sys;

/// The module that contains all events available in the framework.
pub mod events {
    pub use crate::html::{ChangeData, InputData};

    #[doc(no_inline)]
    pub use web_sys::{
        AnimationEvent, DragEvent, ErrorEvent, Event, FocusEvent, InputEvent, KeyboardEvent,
        MouseEvent, PointerEvent, ProgressEvent, TouchEvent, TransitionEvent, UiEvent, WheelEvent,
    };
}

/// Starts an app mounted to a body of the document.
pub fn start_app<COMP>()
where
    COMP: Component,
    COMP::Properties: Default,
{
    App::<COMP>::new().mount_to_body();
}

/// Starts an app mounted to a body of the document.
pub fn start_app_with_props<COMP>(props: COMP::Properties)
where
    COMP: Component,
{
    App::<COMP>::new().mount_to_body_with_props(props);
}

/// The Yew Prelude
///
/// The purpose of this module is to alleviate imports of many common types:
///
/// ```
/// # #![allow(unused_imports)]
/// use yew::prelude::*;
/// ```
pub mod prelude {
    #[cfg(feature = "agent")]
    pub use crate::agent::{Bridge, Bridged, Dispatched, Threaded};
    pub use crate::app::App;
    pub use crate::callback::Callback;
    pub use crate::events::*;
    pub use crate::html::{
        Children, ChildrenWithProps, Classes, Component, ComponentLink, Html, NodeRef, Properties,
        ShouldRender,
    };
    pub use crate::macros::{classes, html, html_nested};

    /// Prelude module for creating worker.
    #[cfg(feature = "agent")]
    pub mod worker {
        pub use crate::agent::{
            Agent, AgentLink, Bridge, Bridged, Context, HandlerId, Job, Private, Public,
        };
    }
}

pub use self::prelude::*;
