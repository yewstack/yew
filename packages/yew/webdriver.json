{
  "goog:chromeOptions": {
    "args": [
      "--use-fake-ui-for-media-stream",
      "--use-fake-device-for-media-stream"
    ]
  }
}
