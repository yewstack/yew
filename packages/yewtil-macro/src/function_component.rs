use proc_macro::TokenStream as TokenStream1;
use proc_macro2::{Ident, Span, TokenStream};
use quote::{quote, ToTokens};
use syn::parse::{Parse, ParseBuffer};
use syn::parse_macro_input;
use syn::punctuated::Punctuated;
use syn::token;
use syn::Token;
use syn::{braced, parenthesized};
use syn::{Block, Error, Field, Stmt, Type, VisPublic, Visibility};

pub fn function_component_handler(attr: TokenStream, item: TokenStream1) -> TokenStream1 {
    let component_name = attr.to_string();
    assert!(
        !component_name.is_empty(),
        "you must provide a component name. eg: function_component(MyComponent)"
    );
    let component_name = Ident::new(&component_name, Span::call_site());
    let function = parse_macro_input!(item as Function);
    TokenStream1::from(
        FunctionComponentInfo {
            component_name,
            function,
        }
        .to_token_stream(),
    )
}

pub struct FunctionComponentInfo {
    component_name: Ident,
    function: Function,
}

// TODO, support type parameters

pub struct Function {
    pub vis: Visibility,
    pub fn_token: Token![fn],
    pub name: Ident,
    pub paren_token: token::Paren,
    pub fields: Punctuated<Field, Token![,]>,
    pub returns_token: Token![->],
    pub return_ty: Ident,
    pub brace_token: token::Brace,
    pub body: Vec<Stmt>,
}

impl Parse for Function {
    fn parse(input: &ParseBuffer) -> Result<Self, Error> {
        let vis = input.parse()?;
        let fn_token = input.parse()?;
        let name = input.parse()?;
        let content;
        let paren_token = parenthesized!(content in input);
        let returns_token = input.parse()?;
        let return_ty = input.parse()?;
        let content2;
        let brace_token = braced!(content2 in input);
        Ok(Function {
            vis,
            fn_token,
            name,
            paren_token,
            fields: content.parse_terminated(Field::parse_named)?,
            returns_token,
            return_ty,
            brace_token,
            body: content2.call(Block::parse_within)?,
        })
    }
}

impl ToTokens for Function {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let Function {
            vis,
            fn_token,
            name,
            fields,
            returns_token,
            return_ty,
            body,
            ..
        } = self;
        let fields = fields
            .iter()
            .map(|field: &Field| {
                let mut new_field: Field = field.clone();
                new_field.attrs = vec![];
                new_field
            })
            .collect::<Punctuated<_, Token![,]>>();

        tokens.extend(quote! {
            #vis #fn_token #name(#fields) #returns_token #return_ty {
                #(#body)*
            }
        })
    }
}

impl ToTokens for FunctionComponentInfo {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let FunctionComponentInfo {
            component_name,
            function,
        } = self;
        // The function tokens must be re-generated in order to strip the attributes that are not allowed.
        let function_token_stream = function.to_token_stream();
        let Function {
            vis, name, fields, ..
        } = function;

        let impl_name = format!("FuncComp{}", component_name.to_string());
        let impl_name = Ident::new(&impl_name, Span::call_site());

        let alias = quote! {
            #vis type #component_name = ::yewtil::Pure<#impl_name>;
        };

        // Set the fields to be public and strips references as necessary.
        // This will preserve attributes like #[props(required)], which will appear in the generated struct below.
        let new_fields = fields
            .iter()
            .map(|field: &Field| {
                let mut new_field: Field = field.clone();
                let visibility = Visibility::Public(VisPublic {
                    pub_token: syn::token::Pub {
                        span: Span::call_site(),
                    },
                });
                // Strip references so the component can have a static lifetime.
                // TODO Handle 'static lifetimes gracefully here - allowing &'static strings instead of erroneously converting them to plain strs.
                let ty = match &field.ty {
                    Type::Reference(x) => {
                        let elem = x.elem.clone();
                        Type::Verbatim(quote! {
                            #elem
                        })
                    }
                    x => x.clone(),
                };
                new_field.vis = visibility;
                new_field.ty = ty;
                new_field
            })
            .collect::<Punctuated<_, Token![,]>>();

        let component_struct = quote! {
            #[derive(::std::clone::Clone, ::std::cmp::PartialEq, ::yew::Properties)]
            #vis struct #impl_name {
                #new_fields
            }
        };

        let arguments = fields
            .iter()
            .zip(new_fields.iter())
            .map(|(field, new_field): (&Field, &Field)| {
                let field_name = field.ident.as_ref().expect("Field must have name");

                // If the fields differ, then a reference was removed from the function's field's type
                // to make it static.
                // Otherwise it is assumed that the type is not a reference on the function and it
                // implements clone, and that when calling the function, the type should be cloned again.
                if field.ty != new_field.ty {
                    quote! {
                        &self.#field_name
                    }
                } else {
                    quote! {
                        self.#field_name.clone()
                    }
                }
            })
            .collect::<Punctuated<_, Token![,]>>();

        let pure_component_impl = quote! {
            impl ::yewtil::PureComponent for #impl_name {
                fn render(&self) -> ::yew::Html {
                     #name(#arguments)
                }
            }
        };

        tokens.extend(quote! {
            #function_token_stream
            #alias
            #component_struct
            #pure_component_impl
        })
    }
}
