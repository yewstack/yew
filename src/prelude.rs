//! The Yew Prelude
//!
//! The purpose of this module is to alleviate imports of many common types:
//!
//! ```
//! # #![allow(unused_imports)]
//! use yew::prelude::*;
//! ```
extern crate stdweb;

pub use html::{
    Component,
    ComponentLink,
    Href,
    Html,
    ChangeData,
    InputData,
    Renderable,
    ShouldRender,
};

pub use app::App;

pub use callback::Callback;

pub use stdweb::web::event::{
    BlurEvent,
    ClickEvent,
    DoubleClickEvent,
    IKeyboardEvent,
    IMouseEvent,
    KeyDownEvent,
    KeyPressEvent,
    KeyUpEvent,
    MouseDownEvent,
    MouseMoveEvent,
    MouseOutEvent,
    MouseOverEvent,
    MouseUpEvent,
};

pub use agent::{
    Ambit,
    Bridge,
    Worker,
};

/// Prelude module for creating worker.
pub mod worker {
    pub use agent::{
        Addr,
        Agent,
        AgentLink,
        HandlerId,
        Transferable,
        Worker,
    };
}
