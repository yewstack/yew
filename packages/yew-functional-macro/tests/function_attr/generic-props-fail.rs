use yew::prelude::*;
use yew_functional::function_component;

#[derive(Clone, Properties, PartialEq)]
struct Props {
    a: usize,
}

#[function_component(Comp)]
fn comp<P: Properties + PartialEq>(_props: &P) -> Html {
    html! {
        <p></p>
    }
}

fn compile_fail() {
    // missing prop 'a'
    html! { <Comp<Props> /> };
    
    // invalid type parameter
    html! { <Comp<INVALID> /> };
    // parameter doesn't match bounds
    html! { <Comp<()> /> };

    // missing type param
    html! { <Comp /> };
}


fn main() {}
