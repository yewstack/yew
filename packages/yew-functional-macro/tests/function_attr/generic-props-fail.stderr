error[E0412]: cannot find type `INVALID` in this scope
  --> $DIR/generic-props-fail.rs:26:19
   |
21 | fn compile_fail() {
   |                - help: you might be missing a type parameter: `<INVALID>`
...
26 |     html! { <Comp<INVALID> /> };
   |                   ^^^^^^^ not found in this scope

error[E0599]: no method named `build` found for struct `PropsBuilder<PropsBuilderStep_missing_required_prop_a>` in the current scope
  --> $DIR/generic-props-fail.rs:23:14
   |
4  | #[derive(Clone, Properties, PartialEq)]
   |                 ---------- method `build` not found for this
...
23 |     html! { <Comp<Props> /> };
   |              ^^^^ method not found in `PropsBuilder<PropsBuilderStep_missing_required_prop_a>`

error[E0599]: the function or associated item `new` exists for struct `VChild<FunctionComponent<comp<MissingTypeBounds>>>`, but its trait bounds were not satisfied
  --> $DIR/generic-props-fail.rs:28:14
   |
28 |     html! { <Comp<MissingTypeBounds> /> };
   |              ^^^^ function or associated item cannot be called on `VChild<FunctionComponent<comp<MissingTypeBounds>>>` due to unsatisfied trait bounds
   |
  ::: $WORKSPACE/packages/yew-functional/src/lib.rs
   |
   | pub struct FunctionComponent<T: FunctionProvider + 'static> {
   | ----------------------------------------------------------- doesn't satisfy `_: yew::Component`
   |
   = note: the following trait bounds were not satisfied:
           `FunctionComponent<comp<MissingTypeBounds>>: yew::Component`

error[E0277]: the trait bound `MissingTypeBounds: yew::Properties` is not satisfied
  --> $DIR/generic-props-fail.rs:28:14
   |
28 |     html! { <Comp<MissingTypeBounds> /> };
   |              ^^^^ the trait `yew::Properties` is not implemented for `MissingTypeBounds`
   |
   = note: required because of the requirements on the impl of `FunctionProvider` for `comp<MissingTypeBounds>`

error[E0277]: can't compare `MissingTypeBounds` with `MissingTypeBounds`
  --> $DIR/generic-props-fail.rs:28:14
   |
28 |     html! { <Comp<MissingTypeBounds> /> };
   |              ^^^^ no implementation for `MissingTypeBounds == MissingTypeBounds`
   |
   = help: the trait `PartialEq` is not implemented for `MissingTypeBounds`
   = note: required because of the requirements on the impl of `FunctionProvider` for `comp<MissingTypeBounds>`

error[E0107]: missing generics for type alias `Comp`
  --> $DIR/generic-props-fail.rs:31:14
   |
31 |     html! { <Comp /> };
   |              ^^^^ expected 1 type argument
   |
note: type alias defined here, with 1 type parameter: `P`
  --> $DIR/generic-props-fail.rs:9:22
   |
9  | #[function_component(Comp)]
   |                      ^^^^
10 | fn comp<P>(_props: &P) -> Html
   |         -
help: use angle brackets to add missing type argument
   |
31 |     html! { <Comp<P> /> };
   |                  ^^^
